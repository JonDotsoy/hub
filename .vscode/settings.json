{
  "cSpell.words": ["dtos"],
  "yaml.schemas": {
    "file:///Users/jondotsoy/Repositories/jondotsoy/q/schema.json": [
      ".agent",
      "*.agent",
      ".cofy",
      "*.cofy"
    ],
    "file:///Users/jonathan.delgado/.vscode/extensions/atlassian.atlascode-3.0.10/resources/schemas/pipelines-schema.json": "bitbucket-pipelines.yml",
    "https://spec.openapis.org/oas/3.1/schema/2022-10-07": "file:///Users/jonathan.delgado/Repositories/jondotsoy/ondina/packages/hub/src/service/openapi.yml"
  }
}
