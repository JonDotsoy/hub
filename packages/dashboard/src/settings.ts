import { get } from "@jondotsoy/utils-js/get"

export const settings = {
  hub: {
    url: get.string(process.env, "HUB_URL"),
  }
}
