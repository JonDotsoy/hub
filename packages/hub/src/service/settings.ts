import { get } from "@jondotsoy/utils-js/get";
import * as YAML from "yaml";

namespace loadHubSchema {
  const downloadHttpAgent = async (target: URL) => (await fetch(target)).text();
  const downloadFileAgent = async (target: URL) => Bun.file(target).text();

  const downloadAgents: Record<
    string,
    undefined | ((target: URL) => Promise<string>)
  > = {
    "http:": downloadHttpAgent,
    "https:": downloadHttpAgent,
    "file:": downloadFileAgent,
  };

  const parsers: Record<string, undefined | ((payload: string) => any)> = {
    ".yml": YAML.parse,
    ".yaml": YAML.parse,
    ".json": JSON.parse,
  };

  const downloadSources = (target: URL) => {
    const agent = downloadAgents[target.protocol];
    if (!agent) throw new Error(`Unsupported protocol ${target.protocol}`);
    return agent(target);
  };

  const parseSource = (target: URL, source: string) => {
    const getParse = () => {
      for (const [name, parse] of Object.entries(parsers)) {
        if (target.pathname.endsWith(name)) return parse;
      }
      return null;
    };

    const parse = getParse();
    if (!parse) throw new Error(`Unsupported extension ${target.pathname}`);
    return parse(source);
  };

  export const load = async (target: string) => {
    const targetUrl = new URL(target, `file:${process.cwd()}/`);
    const source = await downloadSources(targetUrl);
    const data = parseSource(targetUrl, source);
    return data;
  };
}

const getStringNumber = (obj: unknown, ...paths: PropertyKey[]) => {
  const value = get(obj, ...paths);
  if (typeof value === "number") return value;
  if (typeof value === "string" && /^\d+(\.\d+)?$/.test(value))
    return Number(value);
  return null;
};

const getHubSchema = async (
  obj: unknown,
  ...paths: PropertyKey[]
): Promise<unknown> => {
  const hubManifestLocation = get.string(obj, ...paths);
  if (hubManifestLocation) {
    return await loadHubSchema.load(hubManifestLocation);
  }

  return null;
};

export const loadSettings = async (
  envs: Record<string, string | undefined> = process.env,
) => {
  return {
    port: getStringNumber(envs, "PORT") ?? 3000,
    base: get.string(envs, "BASE") ?? "/",
    site: get.string(envs, "SITE"),
    hubSchema: await getHubSchema(envs, "MANIFEST_LOCATION"),
  };
};

export const settings = await loadSettings(process.env);
