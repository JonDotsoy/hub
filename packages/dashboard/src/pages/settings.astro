---
import DefaultLayout from "../components/common/layaouts/default.astro";
import { settings } from "../settings.js";

export const prerender = false;
---

<DefaultLayout title="Settings">
  <pre><code>{JSON.stringify(settings,null,2)}</code></pre>
</DefaultLayout>
