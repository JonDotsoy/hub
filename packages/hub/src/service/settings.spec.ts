import { describe, it, beforeAll, expect } from "bun:test";
import * as fs from "fs/promises";
import { loadSettings } from "./settings.js";

describe("loadSettings", () => {
  it("loads settings from a YAML manifest", async () => {
    const manifestPath = new URL(
      "__samples__/manifest_formats/manifest.yaml",
      import.meta.url,
    );

    const settings = await loadSettings({
      MANIFEST_LOCATION: manifestPath.pathname.toString(),
    });

    expect(settings.hubSchema?.get()).toMatchSnapshot();
  });

  it("loads settings from a JSON manifest", async () => {
    const manifestPath = new URL(
      "__samples__/manifest_formats/manifest.json",
      import.meta.url,
    );

    const settings = await loadSettings({
      MANIFEST_LOCATION: manifestPath.pathname.toString(),
    });

    expect(settings.hubSchema?.get()).toMatchSnapshot();
  });
});
