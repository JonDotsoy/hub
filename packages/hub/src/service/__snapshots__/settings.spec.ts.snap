// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`loadSettings loads settings from a YAML manifest 1`] = `
{
  "permissions": [
    "users.list",
    "users.create",
    "users.delete",
    "sales.read",
    "sales.write",
  ],
  "principals": [
    {
      "id": "bob",
      "roles": [
        "rrhh",
      ],
    },
    {
      "id": "alice",
      "roles": [
        {
          "condition": {
            "equal": [
              "group.office",
              "NY",
            ],
          },
          "role": "admin",
        },
      ],
    },
  ],
  "roles": [
    {
      "id": "rrhh",
      "permissions": [
        "users.list",
        "users.create",
      ],
    },
    {
      "id": "admin",
      "permissions": [
        "users.list",
        "users.create",
        "users.delete",
      ],
    },
  ],
}
`;

exports[`loadSettings loads settings from a JSON manifest 1`] = `
{
  "permissions": [
    "users.list",
    "users.create",
    "users.delete",
    "sales.read",
    "sales.write",
  ],
  "principals": [
    {
      "id": "bob",
      "roles": [
        "rrhh",
      ],
    },
    {
      "id": "alice",
      "roles": [
        {
          "condition": {
            "equal": [
              "group.office",
              "NY",
            ],
          },
          "role": "admin",
        },
      ],
    },
  ],
  "roles": [
    {
      "id": "rrhh",
      "permissions": [
        "users.list",
        "users.create",
      ],
    },
    {
      "id": "admin",
      "permissions": [
        "users.list",
        "users.create",
        "users.delete",
      ],
    },
  ],
}
`;
